spec/offline_sort/offline_sort_spec.rb

```diff
@@ -1,113 +1,128 @@
-      OfflineSort.sort(enumerable, chunk_size: entries_per_chunk, &sort)
+    let(:offline_sorter_instance) do
+      described_class.new(enumerable, chunk_size: entries_per_chunk, &sort)
     end
 
-    it "writes out to disk" do
-      expect(Tempfile).to receive(:open).at_least(:once).and_call_original
-      subject
+    let(:entries_per_chunk) { 900 }
+    let(:count) { 10000 }
+    let(:enumerable) { arrays }
+    let(:sort) { array_sort }
+    let(:array_sort) { Proc.new { |arr| arr[2] } }
+    let(:arrays) do
+      count.times.map do |index|
+        [SecureRandom.hex, index, SecureRandom.hex]
+      end
     end
 
-    shared_examples "produces a sorted result" do
-      it "produces the same sorted result as an in-memory sort" do
-        sorted = subject.to_a
-
-        expect(unsorted).to match_array(enumerable)
-        expect do
-          last = nil
-          entry_count = 0
-          sorted.each do |entry|
-            if last.nil?
-              last = entry
-              entry_count += 1
-              next
-            end
+    shared_examples "a correct offline sort" do
+      let(:unsorted) { enumerable.dup }
 
-            unless (sort.call(last) <=> sort.call(entry)) == -1
-              raise "Out of order at line #{entry_count}"
+      it "writes out to disk" do
+        expect(Tempfile).to receive(:open).at_least(:once).and_call_original
+        subject
+      end
+
+      shared_examples "produces a sorted result" do
+        it "produces the same sorted result as an in-memory sort" do
+          expect(unsorted.sort_by(&sort)).to eq(subject)
+        end
+
+        context "closing tempfiles" do
+          it "closes all tempfiles" do
+            close_count = 0
+            allow_any_instance_of(Tempfile).
+              to receive(:close) { close_count += 1 }
+
+            expected_number_of_tempfiles =
+              (count.to_f / entries_per_chunk).ceil
+
+            # The case where we don't write to disk because there's only
+            # one chunk
+            if expected_number_of_tempfiles == 1
+              expected_number_of_tempfiles = 0
             end
 
-            last = entry
-            entry_count += 1
+            subject
+            expect(close_count).to eq(expected_number_of_tempfiles)
           end
-        end.not_to raise_error
-        expect(sorted).to match_array(enumerable.sort_by(&sort))
-      end
-    end
+        end
 
-    context "when the number of entries is smaller than the chunk size" do
-      let(:count) { entries_per_chunk - 1 }
+        context "when sorted twice" do
+          subject { offline_sorter_instance.sort }
 
-      it "does not write out to disk" do
-        expect(Tempfile).not_to receive(:open)
-        subject
+          it "produces the same result both times" do
+            expect(subject.to_a).to eq(subject.to_a)
+          end
+
+          it "only calls sort the first time" do
+            offline_sort = offline_sorter_instance.sort.to_a
```

Comment:
I think that moving this `to_a` call to happen after the `enumerable` will make this test have a stronger assertion as it'll cover that the state of the enumerable is flushed to something inside fo the `Merger`. With the `to_a` call here happening before the `enumerable.clear` I think we could have this pass even if that weren't the case. Are you seeing that differently though?

Reply:
ALso could you add some comments explaining the reasoning for the ordering between the sort and clear calls? I think things are a little intricate here and can def see myself forgetting why we laid them out this way.